pkg/cloudprovider/openstack/openstack.go

```diff
@@ -651,45 +696,53 @@ func (lb *LoadBalancer) UpdateTCPLoadBalancer(name, region string, hosts []strin
 func (lb *LoadBalancer) EnsureTCPLoadBalancerDeleted(name, region string) error {
 	glog.V(4).Infof("EnsureTCPLoadBalancerDeleted(%v, %v)", name, region)
 
-	// TODO(#8352): Because we look up the pool using the VIP object, if the VIP
-	// is already gone we can't attempt to delete the pool. We should instead
-	// continue even if the VIP doesn't exist and attempt to delete the pool by
-	// name.
-	vip, vipErr := getVipByName(lb.network, name)
-	if vipErr == ErrNotFound {
-		return nil
-	} else if vipErr != nil {
-		return vipErr
-	}
-
-	// It's ok if the pool doesn't exist, as we may still need to delete the vip
-	// (although I don't believe the system should ever be in that state).
-	pool, poolErr := pools.Get(lb.network, vip.PoolID).Extract()
-	if poolErr != nil {
-		detailedErr, ok := poolErr.(*gophercloud.UnexpectedResponseCodeError)
-		if !ok || detailedErr.Actual != http.StatusNotFound {
-			return poolErr
-		}
-	}
-	poolExists := (poolErr == nil)
-
-	// We have to delete the VIP before the pool can be deleted, so we can't
-	// continue on if this fails.
-	// TODO(#8352): Only do this if the VIP exists once we can delete pools by
-	// name rather than by ID.
-	err := vips.Delete(lb.network, vip.ID).ExtractErr()
+	vip, err := getVipByName(lb.network, name)
 	if err != nil && err != ErrNotFound {
 		return err
 	}
 
-	// Ignore errors for everything following here
+	// We have to delete the VIP before the pool can be deleted,
+	// so no point continuing if this fails.
+	if vip != nil {
+		err := vips.Delete(lb.network, vip.ID).ExtractErr()
+		if err != nil && !isNotFound(err) {
+			return err
+		}
+	}
 
-	if poolExists {
+	var pool *pools.Pool
+	if vip != nil {
+		pool, err = pools.Get(lb.network, vip.PoolID).Extract()
+		if err != nil && !isNotFound(err) {
+			return err
+		}
+	} else {
+		// The VIP is gone, but it is conceivable that a Pool
+		// still exists that we failed to delete on some
+		// previous occasion.  Make a best effort attempt to
+		// cleanup any pools with the same name as the VIP.
+		pool, err = getPoolByName(lb.network, name)
+		if err != nil && err != ErrNotFound {
+			return err
+		}
+	}
+
+	if pool != nil {
 		for _, monId := range pool.MonitorIDs {
-			// TODO(#8352): Delete the monitor, don't just disassociate it.
-			pools.DisassociateMonitor(lb.network, pool.ID, monId)
+			_, err = pools.DisassociateMonitor(lb.network, pool.ID, monId).Extract()
+			if err != nil {
+				return err
+			}
+
+			err = monitors.Delete(lb.network, monId).ExtractErr()
+			if err != nil {
```

Comment:
Probably not needed since the retry would succeed, but if some parallel process were also messing with the resources then this could conceivably be a not-found error.